### PR TITLE
29_Xinpeng-Huang Homework 3

### PR DESCRIPTION
**Pull request by xph**

### CODE REVIEW

File: Lesson3/assignment/yours.sol [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+=> [B, O]
+contract C is O
+=> [C, O]
+contract K1 is A, B
+=> K1 + merge(L[B], L[A], [B, A])
+   K1 + merge([B, O], [A, O], [B, A])
+   [K1, B, A, O]
+contract K2 is A, C
+=> [K2, C, A, O]
+contract Z is K1, K2
```
Comment (jonnyhsy) [inline Lesson3/assignment/yours.sol]: 前面几步都正确，最后一步：
L(Z) = [Z] + merge(L(K2), L(K1), [K2, K1])
     = [Z] + merge([K2, C, A, O], [K1, B, A, O], [K2, K1])
     = [Z, K2] + merge([C, A, O], [K1, B, A, O], [K1])
     = [Z, K2, C] + merge([A, O], [K1, B, A, O], [K1])
     = [Z, K2, C, K1] + merge([A, O], [B, A, O])
     = [Z, K2, C, K1, B] + merge([A, O], [A, O])
     = [Z, K2, C, K1, B, A, O]

File: Lesson3/assignment/yours.sol [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        assert(employee.id != 0x0);
+        _;
+    }
+    
+    function Payroll() {
+        owner = msg.sender;
+    }
+    
+    function _partialPaid(Employee employee) private {
+        uint partialPayment = employee.salary * (now - employee.lastPayday) / payDuration;
```
Comment (jonnyhsy) [inline Lesson3/assignment/yours.sol]: 设计数学运算的地方，建议使用课程里提及的OpenZepplin的SafeMath.sol库来防止溢出，可以在开头导入：
```
import "github.com/OpenZeppelin/zeppelin-solidity/contracts/math/SafeMath.sol";
import "github.com/OpenZeppelin/zeppelin-solidity/contracts/ownership/Ownable.sol";
```
然后
在contract里：using SafeMath for uint;
在此处： uint partialPayment = employee.salary.mul(now.sub(employee.lastPayday)).div(payDuration);
参见文档：
 - https://solidity.readthedocs.io/en/develop/contracts.html#using-for
 - https://github.com/OpenZeppelin/zeppelin-solidity/blob/c7f2138da1b5d588aa1b93b6201b0272f03f3661/contracts/math/SafeMath.sol

File: Lesson3/assignment/yours.sol [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        uint lastPayday;
+    }
+    
+    uint constant payDuration = 10 seconds;
+    uint totalSalary;
+
+    address owner;
+    mapping(address => Employee) public employees;
+    
+    modifier onlyOwner {
```
Comment (jonnyhsy) [inline Lesson3/assignment/yours.sol]: 可以重用openzepplin的Ownable.sol代码:
https://github.com/OpenZeppelin/zeppelin-solidity/blob/c7f2138da1b5d588aa1b93b6201b0272f03f3661/contracts/ownership/Ownable.sol

File: Lesson3/assignment/yours.sol (hunk 1 above)
Comment (xph) [reply Lesson3/assignment/yours.sol]: 哦，错看了那个C，谢谢纠正！

File: Lesson3/assignment/yours.sol (hunk 2 above)
Comment (xph) [reply Lesson3/assignment/yours.sol]: 谢谢提醒，这一步忘了，我补上

**Comment by xph** [comment]
@jonnyhsy 改好了

**Pull request merged**